### PR TITLE
TeamCity-Kobiton

### PR DESCRIPTION
**Pull request by phanthaoIT**

### CODE REVIEW

File: integrate-kobiton-teamcity.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  + [1.Preparing automation test repository](#1-preparing-automation-test-repository)
+  + [2. Getting desired capabilities](#2-getting-desired-capabilities)
+  + [3. Configuring automation test script](#3-configuring-automation-test-script)
+  + [4. Configuring TeamCity project](#4-configuring-teamcity-project)
++ [C. Execution](#c-execution)
+  + [1. Runnig automation test on Kobiton devices](#1-running-automation-test-on-kobiton-devices)
+  + [2. Fetch test session data through Kobiton REST API](#2-fetch-test-session-data-through-kobiton-rest-api)
++ [D. Feedback](#d-feedback)
+## A. PREREQUISTIES 
+- Kobiton account: Go to https://portal.kobiton.com/register to create new account.
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: `Go to https://portal.kobiton.com/register to create new account` -> `Go to https://portal.kobiton.com/register to create a new account if you have not created before.`

File: integrate-kobiton-teamcity.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+- Kobiton Username and API Key: Folow instruction at `IV. Configure Test Script for Kobiton` section on [our blog article](https://kobiton.com/blog/tutorial/parallel-testing-selenium-webdriver/) to get Username and API Key. 
+## B. SETUP
+### 1. Preparing automation test repository
+To perform automation testing on Kobiton, you need a GitHub repository fully configured with TeamCity. We have provided a sample script for executing automation test with Kobiton.
+
+To use the provided sample, follow these step:
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: `To use the provided sample, follow these step:` -> `To use the provided sample, follow these steps:`

File: integrate-kobiton-teamcity.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
++ Clone [this repository](https://github.com/kobiton/TeamCity-HockeyApp-Appium-ReactNative).
++ Create a new GitHub repository.
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: What does the GitHub repo use for? Please describe a little bit.

File: integrate-kobiton-teamcity.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
++ Copy contents of `sample` folder into your repository and create a project in TeamCity with this repository URL.
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: You used conjuration `and`, it means there are 2 steps to do. Please break it into 2 steps:
* Copy contents of `sample` folder into your repository.
* Create and configure a project in TeamCity to use this repo.

Comment (baotn166) [reply integrate-kobiton-teamcity.md]: How can I create and configure TeamCity? If you have written the way to configure, please provide the link to the content here.

File: integrate-kobiton-teamcity.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  deviceOrientation:  'portrait',  
+  captureScreenshots: true, 
+  app:                '<APP_URL>', 
+  deviceGroup:        'KOBITON', 
+  deviceName:         '<DEVICE_NAME>',
+  platformVersion:    '<DEVICE_VERSION>',
+  platformName:       '<DEVICE_PLATFORM_NAME>' 
+}
+```
+
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: We should describe a little bit parameters (which is required or not):
* YOUR_APP_URL: (app repo or external link)
* DEVICE_NAME
* PLATFORM_VERSION
* PLATFORM_NAME

You should provide some more documentation to describe clearer about parameters, like [this](https://github.com/kobiton/CircleCI-HockeyApp-Appium-ReactNative/blob/0406b8d1f1d94dd2e8f9367e22b3ac16a7678c1d/Integrate-Kobiton.md#22-setting-desired-capabilities).

File: integrate-kobiton-teamcity.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+### 4. Configuring TeamCity project
+**Attaching environment variable to Teamcity**
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: `Attaching environment variable to Teamcity` -> `Attaching environment variable to TeamCity`

File: integrate-kobiton-teamcity.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+Add your Kobiton username and API key as environment variables.
+
+ ![env](./assets/env.png)
+
+**Setup TeamCity automation testing**
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: `Setup TeamCity automation testing` -> `Setup TeamCity to run automation test`

File: integrate-kobiton-teamcity.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+In your project configuration, add a build step with `Command Line` as runner type and custom scripts.
+
+Example: 
+
+![command](./assets/cmd.png)
+
+## C. EXECUTION
+### 1. Running automation test on Kobiton devices
++ Push your changes to GitHub. TeamCity will install necessary dependencies and then execute the test on Kobiton.
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: What changes should I have to do?

**Review by baotn166 — CHANGES_REQUESTED** [review]
@phanthaoIT Please consider my comments and fix if those make sense.

File: integrate-kobiton-teamcity.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
++ [B. Setup](#b-setup)
+  + [1.Preparing automation test repository](#1-preparing-automation-test-repository)
+  + [2. Getting desired capabilities](#2-getting-desired-capabilities)
+  + [3. Configuring automation test script](#3-configuring-automation-test-script)
+  + [4. Configuring TeamCity project](#4-configuring-teamcity-project)
++ [C. Execution](#c-execution)
+  + [1. Runnig automation test on Kobiton devices](#1-running-automation-test-on-kobiton-devices)
+  + [2. Fetch test session data through Kobiton REST API](#2-fetch-test-session-data-through-kobiton-rest-api)
++ [D. Feedback](#d-feedback)
+## A. PREREQUISTIES
```
Comment (baotn166) [inline integrate-kobiton-teamcity.md]: `PREREQUISTIES ` --> `PREREQUISITES`

**Review by baotn166 — CHANGES_REQUESTED** [review]

File: Readme/integrate-kobiton-teamcity.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  + [3. Configuring TeamCity project](#3-configuring-teamcity-project)
++ [C. Execution](#c-execution)
+  + [1. Runnig automation test on Kobiton devices](#1-running-automation-test-on-kobiton-devices)
+  + [2. Fetch test session data through Kobiton REST API](#2-fetch-test-session-data-through-kobiton-rest-api)
++ [D. Feedback](#d-feedback)
+## A. PREPARATION
+### 1. Gettting Kobiton username and API key
+In order to execute test on Kobiton, these two must-have parameters must be provided: Username and API key. They are necessary for authenticating with Kobiton.
+
+If you don't have a Kobiton account, visit https://portal.kobiton.com/register to create an account.
```
Comment (tiennguyenhoan) [inline Readme/integrate-kobiton-teamcity.md]:
```
>If you don't have a Kobiton account, visit https://portal.kobiton.com/register to create an account.

File: Readme/integrate-kobiton-teamcity.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+Follow the instruction at `IV. Configure Test Script for Kobiton` section on [our blog article](https://kobiton.com/blog/tutorial/parallel-testing-selenium-webdriver/) to get Kobiton username and API key.
+
+### 2. Sample
+In this repository, we have already provided samples for executing automation test with Kobiton through TeamCity:
+  -  [ios-app-test.js](../sample/automation-script/ios-app-test.js): script for execute automation test on Kobiton iOS devices.
+  -  [android-app-test.js](../sample/automation-script/android-app-test.js): script for execute automation test on Kobiton Android devices.
+
+To use the provided sample, follow these steps:
+1. Fork [this repository](https://github.com/kobiton/TeamCity-HockeyApp-Appium-ReactNative).
```
Comment (tiennguyenhoan) [inline Readme/integrate-kobiton-teamcity.md]:
```
1. Fork this repository: https://github.com/kobiton/TeamCity-HockeyApp-Appium-ReactNative
```

File: Readme/integrate-kobiton-teamcity.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+2. Create a project in TeamCity to use the forked repository. If you don't know how to do it, please follow [TeamCity document](https://confluence.jetbrains.com/display/TCD9/Creating+and+Editing+Projects#CreatingandEditingProjects-CreateProjectfromURL) for instructions.
```
Comment (tiennguyenhoan) [inline Readme/integrate-kobiton-teamcity.md]: What happens if I already has a project which was used for build and deploy app, do I need to create another one or can I add this repo to the created project

File: Readme/integrate-kobiton-teamcity.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    ![click auto](./assets/click_auto.png)
+ 
+  + On the left-hand side, you can select your preferred language, App Type, etc.
+  In this example, we use **NodeJS** as the default language of the script. Therefore, choose **NodeJS** in `Language` section and **Hybrid/Native from Apps** in `App type` section. On the right-hand side, Kobiton automatically generates your desired capabilities into code based on the selected options.
+
+  + Write down the generated code as we will be using it in the next step.
+
+    ![click auto](./assets/automation.png) 
+### 2. Configuring automation test script
+In the `automation-script` folder, we will run test for `ApiDemos-debug` application on Android and `UIKitCatalog-Test-Adhoc` apllication on iOS.
```
Comment (tiennguyenhoan) [inline Readme/integrate-kobiton-teamcity.md]: I think this is a distract information, what is the point of adding this line to the docs

File: Readme/integrate-kobiton-teamcity.md [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  captureScreenshots: true, 
+  app:                '<YOUR_APP_URL>', 
+  deviceGroup:        'KOBITON', 
+  deviceName:         '<DEVICE_NAME>',
+  platformVersion:    '<PLATFORM_VERSION>',
+  platformName:       '<PLATFORM_NAME>' 
+}
+```
+> More information about automation testing with Kobiton can be found at https://docs.kobiton.com/automation-testing/automation-testing-with-kobiton/.
+### 3. Configuring TeamCity project
```
Comment (tiennguyenhoan) [inline Readme/integrate-kobiton-teamcity.md]: I think this session should be on top

**Review by tiennguyenhoan — CHANGES_REQUESTED** [review]
@phanthaoIT Please take a look !

File: README.md [hunk 15]
```diff
@@ -1,2 +1,115 @@
...
+Go to https://portal.kobiton.com and login to your Kobiton account.
+  1. Click **Devices** in the top navigation bar.
+
+  ![click devices](./docs/assets/devices.png)
+ 
+  2. Hover over any device you want to test with and click on Automation settings button (the one with gear symbol).
+ 
+  ![click auto](./docs/assets/click_auto.png)
+ 
+  3. On the left-hand side, you can select your preferred language, App Type, etc.
```
Comment (tiennguyenhoan) [inline README.md]: Please simplify this part or trying to explain it briefly

File: README.md [hunk 16]
```diff
@@ -1,2 +1,115 @@
...
+Follow the instruction at `IV. Configure Test Script for Kobiton` section on [our blog article](https://kobiton.com/blog/tutorial/parallel-testing-selenium-webdriver/) to get Kobiton username and API key.
+
+### 2. Sample
+In this repository, we have already provided samples for executing automation test with Kobiton through TeamCity:
+  -  [ios-app-test.js](../sample/automation-script/ios-app-test.js): script for execute automation test on Kobiton iOS devices.
+  -  [android-app-test.js](../sample/automation-script/android-app-test.js): script for execute automation test on Kobiton Android devices.
+
+To use the provided sample, follow these steps:
+1. Fork this repository: https://github.com/kobiton/TeamCity-HockeyApp-Appium-ReactNative
+2. In your TeamCity project, create a `Build configurations` to use the forked repository. If you don't know how to do it, please follow [TeamCity document](https://confluence.jetbrains.com/display/TCD18/Build+Configuration+Template) for instructions.
```
Comment (tiennguyenhoan) [inline README.md:29]: Am I need to create a `Build configurations` in the team project ???

**Review by tiennguyenhoan — CHANGES_REQUESTED** [review]
@phanthaoIT please take a look and reference comments from Travis PR to adjust the change

File: README.md [hunk 17]
```diff
@@ -1,2 +1,119 @@
...
+3. Clone the forked repository to configure in the next steps.
+## B. SETUP
+### 1. Getting desired capabilities
+Desired capabilities need to be added to the automation test script to allow the test to be executed on Kobiton device. The desiredCap are just matching only for one device. If you want to test on other devices, please make sure noted the different and customize as your need. 
+
+In the provided sample automation test scripts, we have pre-configured them to execute the provided sample application(s) on one of the available device(s) that has the name `Galaxy` for `Android` and `iPhone` for iOS. If you want to use other specific device(s), follow the instruction below to get the corresponding desired capabilities for that device(s).
+
```
Comment (tiennguyenhoan) [inline README.md:36]: add something like this to make it separate ``` *How to get desired caps*```

File: README.md [hunk 18]
```diff
@@ -1,2 +1,119 @@
...
+Kobiton has already provided two sample applications for testing on Android and iOS:
++ `ApiDemos-debug` app on Android: https://appium.github.io/appium/assets/ApiDemos-debug.apk
++ `UIKitCatalog-Test-Adhoc` app on iOS: https://s3-ap-southeast-1.amazonaws.com/kobiton-devvn/apps-test/UIKitCatalog-Test-Adhoc.ipa
+
+In this tutorial, we will be showing how to get desired capabilities for executing automation test of provided sample application `ApiDemos-debug` on `Pixel 2 XL 8.1.0`.
+  1. Go to https://portal.kobiton.com and login to your Kobiton account.
```
Comment (tiennguyenhoan) [inline README.md]:
```
1. Go to https://portal.kobiton.com and login to your Kobiton account.

2. Click **Devices** in the top navigation bar.

   ![click devices](./docs/assets/devices.png)

3. Hover over any device you want to test with and click on Automation settings button (the one with gear symbol).

  ![click auto](./docs/assets/gear-button.png)

4. On the left-hand side, you can select your preferred language, App Type, etc.

-  In this example, we use **NodeJS** as the default language of the script. Therefore, choose **NodeJS** in `Language` section
-  Select **Hybrid/Native from Url** in `App type` section then pass the provided app url to `Application Url` section. 

Kobiton automatically generates your desired capabilities into code based on the selected options.
```

File: README.md [hunk 19]
```diff
@@ -1,2 +1,119 @@
...
+
+  ![click devices](./docs/assets/devices.png)
+ 
+  3. Hover over any device you want to test with and click on Automation settings button (the one with gear symbol).
+ 
+  ![click auto](./docs/assets/gear-button.png)
+ 
+  4. On the left-hand side, you can select your preferred language, App Type, etc.
+  In this example, we use **NodeJS** as the default language of the script. Therefore, choose **NodeJS** in `Language` section and **Hybrid/Native from Url** in `App type` section. Pass the url we provided below in `Application Url` section. On the right-hand side, Kobiton automatically generates your desired capabilities into code based on the selected options.
+  + Copy the desired capabilities to prepare for the next step.
```
Comment (tiennguyenhoan) [inline README.md]: Remove this line

File: README.md [hunk 20]
```diff
@@ -1,2 +1,119 @@
...
+### 2. Configuring automation test script
+In the `automation-script` folder, we have provided two sample scripts for executing automation test on Kobiton iOS and Android devices.
+
+As mentioned above, the desired capabilities in provided automation test scripts have already been pre-configured. If you have chosen custom desired capabilities, please follow steps below to apply the collected desired capabilities.
+
+1. Open `android-app-test.js` if you want to execute test on Android or open `ios-app-test.js` if you want to execute test on iOS devices.
+
+2. Replace `desiredCaps` in the script with one collected in the previous step.
+
+Example:
```
Comment (tiennguyenhoan) [inline README.md]: ```Example of desired capabilities for executing `ApiDemos-debug` application on `Pixel 2 XL` running `Android 8.1.0`:```

File: README.md [hunk 21]
```diff
@@ -1,2 +1,119 @@
...
+
+ ![env](./docs/assets/env.png)
+
+**Setup TeamCity automation test**
+
+In your project configuration, add a build step with `Command Line` as runner type and custom scripts.
+
+![command](./docs/assets/cmd.png)
+
+If you configured `android-app-test` file, command will look like: `npm run android-app-test` or `npm run ios-app-test` if you configured `ios-app-test`.
```
Comment (tiennguyenhoan) [inline README.md]:
```
Replace the variable `<TEST_SCRIPT>` in execute command as: 
- Android device: 
`npm run android-app-test`

- iOS device:
`npm run ios-app-test`
```

**Review by tiennguyenhoan — CHANGES_REQUESTED** [review]
@phanthaoIT Still have issues

**Pull request closed without merging**